dcp_prototype/backend/wrangling/migrations/utils/migration_utils.py

```diff
@@ -0,0 +1,356 @@
+    """
+
+    if type(source_tuple) != tuple:
+        # convert non tuples to a tuple with one item
+        # this is for consistent data handling
+        process_mapping(source, (source_tuple,), dest_object, dest_attr, mapfunc)
+        return
+
+    if len(source_tuple) > 0:
+        current = source_tuple[0]
+        leftover = source_tuple[1:]
+
+        if current == "*":
+            # "*" signifies that the source is an array, and that we should
+            # process all entries
+            for item in source:
+                process_mapping(item, leftover, dest_object, dest_attr, mapfunc)
+
+        else:
+            item = source[current]
+            process_mapping(item, leftover, dest_object, dest_attr, mapfunc)
+
+    else:
+        mapfunc(dest_object, dest_attr, source)
+
+
+def process_mappings(label, source_data, dest_object, mapping, missing):
+    """Process a collection of mappings"""
+    for source_tuple, dest_attr, mapfunc in mapping:
+        source = source_data
+        try:
+            process_mapping(source, source_tuple, dest_object, dest_attr, mapfunc)
+        except KeyError as e:
+            if source_tuple not in missing:
+                missing[source_tuple] = (label, dest_attr, e)
+
+
+class DatasetMetadata(object):
+    """This class performs the mapping from dcp1 to the new artifact schema.
+    All the dcp1 data object are first added with the add_entity function.
+    When all data is gathered, the process function is called.
+    The data dictionary can then be retrieved with to_dict.
+    """
+
+    def __init__(self):
+        self.artifact = MetadataArtifact()
+        self.entity_data = {}
+        self.missing = {}
+        self.lock = threading.Lock()
+
+    def validate(self, data):
+        """Validate the data with the current artifact schema"""
+        try:
+            import dcp_prototype
+
+            schema_file = pkg_resources.resource_filename(dcp_prototype.__name__, "backend/v0.0.0.json")
+            schema_data = open(schema_file).read()
+            schema = json.loads(schema_data)
+        except FileNotFoundError:
+            # FIXME, the schema is not yet committed to main.
+            # validate should fail if the data cannot be validated.
+            print(f"Warning, the schema could not be validates: {schema_file}")
+            return True
+
+        try:
+            jsonschema.validate(data, schema)
+            return True
+        except jsonschema.ValidationError as e:
+            print("validation error: ", e)
+            return False
+
+    def add_entity(self, file_source, object_data):
+        """Add a dcp1 entity to the entity_data dict"""
+        with self.lock:
+            entity_type = get_entity_type(file_source, object_data)
+            self.entity_data.setdefault(entity_type, []).append(object_data)
+
+    def process_project(self):
+        """Process the one and only project entityyy"""
```

Comment:
Three y's here?

Reply:
Yeah, that's a vi fail :).   "yy" yanks the current line, but only if you'r in the command mode.